src/basicfuns.jl

```diff
@@ -259,12 +254,38 @@ $(SIGNATURES)
 Return the [`softmax transformation`](https://en.wikipedia.org/wiki/Softmax_function)
 applied to `x` *in place*.
 """
-softmax!(x::AbstractArray{<:AbstractFloat}) = softmax!(x, x)
+softmax!(x::AbstractArray{<:Real}; dims=:) = softmax!(x, x; dims=dims)
 
 """
 $(SIGNATURES)
 
 Return the [`softmax transformation`](https://en.wikipedia.org/wiki/Softmax_function)
 applied to `x`.
 """
-softmax(x::AbstractArray{<:Real}) = softmax!(similar(x, float(eltype(x))), x)
+function softmax(x::AbstractArray{<:Real}; dims=:)
+    return softmax!(similar(x, float(eltype(x))), x; dims=dims)
+end
+
+function _softmax!(r, x, ::Colon)
+    length(r) == length(x) || throw(DimensionMismatch("inconsistent array lengths"))
+    u = maximum(x)
+    map!(r, x) do xi
+        return exp(xi - u)
+    end
+    LinearAlgebra.lmul!(inv(sum(r)), r)
+    return r
+end
+
+function _softmax!(r, x, dims)
+    size(r) == size(x) || throw(DimensionMismatch("inconsistent array sizes"))
+    u = maximum(x; dims=dims)
+    r .= exp.(x .- u)
+    if u isa Array{eltype(r)}
+        # array can be reused
+        sum!(u, r)
```

Comment:
Are we sure that `sum(r, dims=dims)` will always use the same eltype as `r`?

Reply:
I thought it was a reasonable assumption since `eltype(r)` has to be a floating point number type. However, we could also remove this optimization or restrict it to a fixed set of types such as e.g. `Array{<:Union{Float64,Float32,Float16}}`.

Reply:
What do you think, should it be removed completely or changed to e.g. `if eltype(r) isa Union{Float16,Float32,Float64} && u isa Array{eltype(r)}`?